// @vendor/library/index.js
System.register("@vendor/library/comps/base", [], function (exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var ComponentBase;
    return {
        setters: [],
        execute: function () {
            ComponentBase = (function () {
                function ComponentBase() {
                }
                ComponentBase.prototype.base = function (input) {
                    return { base: input };
                };
                return ComponentBase;
            }());
            exports_1("ComponentBase", ComponentBase);
        }
    };
});
System.register("@vendor/library/comps/one", ["tslib", "@vendor/library/comps/base"], function (exports_2, context_2) {
    "use strict";
    var __moduleName = context_2 && context_2.id;
    var tslib_1, base_1, ComponentOne;
    return {
        setters: [
            function (tslib_1_1) {
                tslib_1 = tslib_1_1;
            },
            function (base_1_1) {
                base_1 = base_1_1;
            }
        ],
        execute: function () {
            ComponentOne = (function (_super) {
                tslib_1.__extends(ComponentOne, _super);
                function ComponentOne() {
                    return _super !== null && _super.apply(this, arguments) || this;
                }
                ComponentOne.prototype.one = function () {
                    return Object.assign(this.base("one"), { one: "self" });
                };
                return ComponentOne;
            }(base_1.ComponentBase));
            exports_2("ComponentOne", ComponentOne);
        }
    };
});
System.register("@vendor/library/comps/two", ["tslib", "@vendor/library/comps/base"], function (exports_3, context_3) {
    "use strict";
    var __moduleName = context_3 && context_3.id;
    var tslib_2, base_2, ComponentTwo;
    return {
        setters: [
            function (tslib_2_1) {
                tslib_2 = tslib_2_1;
            },
            function (base_2_1) {
                base_2 = base_2_1;
            }
        ],
        execute: function () {
            ComponentTwo = (function (_super) {
                tslib_2.__extends(ComponentTwo, _super);
                function ComponentTwo() {
                    return _super !== null && _super.apply(this, arguments) || this;
                }
                ComponentTwo.prototype.two = function () {
                    return Object.assign(this.base("two"), { two: "self" });
                };
                return ComponentTwo;
            }(base_2.ComponentBase));
            exports_3("ComponentTwo", ComponentTwo);
        }
    };
});
System.register("@vendor/library/index", ["@vendor/library/comps/base", "@vendor/library/comps/one", "@vendor/library/comps/two"], function (exports_4, context_4) {
    "use strict";
    var __moduleName = context_4 && context_4.id;
    function exportStar_1(m) {
        var exports = {};
        for (var n in m) {
            if (n !== "default") exports[n] = m[n];
        }
        exports_4(exports);
    }
    return {
        setters: [
            function (base_3_1) {
                exportStar_1(base_3_1);
            },
            function (one_1_1) {
                exportStar_1(one_1_1);
            },
            function (two_1_1) {
                exportStar_1(two_1_1);
            }
        ],
        execute: function () {
        }
    };
});
